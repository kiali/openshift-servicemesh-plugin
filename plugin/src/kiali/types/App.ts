import { Namespace } from './Namespace';
import { Runtime } from './Workload';
import { AppHealthResponse } from '../types/Health';

export interface AppId {
  cluster?: string;
  namespace: string;
  app: string;
}

export interface AppWorkload {
  workloadName: string;
  istioSidecar: boolean;
  istioAmbient: boolean;
  serviceAccountNames: string[];
  labels: { [key: string]: string };
}

export interface App {
  cluster?: string;
  namespace: Namespace;
  name: string;
  workloads: AppWorkload[];
  serviceNames: string[];
  runtimes: Runtime[];
  health: AppHealthResponse;
}
