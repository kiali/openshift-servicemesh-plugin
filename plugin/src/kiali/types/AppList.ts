import { Namespace } from './Namespace';
import { AppHealth } from './Health';
import { ObjectReference } from './IstioObjects';

export interface AppList {
  namespace: Namespace;
  applications: AppOverview[];
}

export interface AppOverview {
  name: string;
  cluster?: string;
  istioSidecar: boolean;
  istioAmbient: boolean;
  labels: { [key: string]: string };
  istioReferences: ObjectReference[];
  health: AppHealth;
}

export interface AppListItem extends AppOverview {
  namespace: string;
}
