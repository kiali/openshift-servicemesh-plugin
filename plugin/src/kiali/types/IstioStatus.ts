import { t } from 'utils/I18nUtils';

export enum Status {
  Healthy = 'Healthy',
  Unhealthy = 'Unhealthy',
  Unreachable = 'Unreachable',
  NotFound = 'NotFound',
  NotReady = 'NotReady'
}

export const statusMsg = {
  [Status.Healthy]: t('Healthy'),
  [Status.NotFound]: t('Not found'),
  [Status.NotReady]: t('Not ready'),
  [Status.Unhealthy]: t('Not healthy'),
  [Status.Unreachable]: t('Unreachable')
};

export interface ComponentStatus {
  is_core: boolean;
  name: string;
  status: Status;
}

export interface IstiodResourceThresholds {
  cpu: number;
  memory: number;
}
