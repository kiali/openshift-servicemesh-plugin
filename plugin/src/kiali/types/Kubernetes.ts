export interface TypeMeta {
  apiVersion: string;
  kind: string;
}
