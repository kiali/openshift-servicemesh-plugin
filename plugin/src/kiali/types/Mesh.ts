import { Controller, ElementModel, GraphElement } from '@patternfly/react-topology';
import { AppenderString } from './Common';

export interface MeshCluster {
  accessible: boolean;
  apiEndpoint: string;
  isKialiHome: boolean;
  kialiInstances: KialiInstance[];
  name: string;
  network: string;
  secretName: string;
}

export interface KialiInstance {
  namespace: string;
  operatorResource: string;
  serviceName: string;
  url: string;
  version: string;
}

export type MeshClusters = MeshCluster[];

// MESH GRAPH

export enum MeshInfraType {
  CLUSTER = 'cluster',
  GRAFANA = 'grafana',
  ISTIOD = 'istiod',
  KIALI = 'kiali',
  METRIC_STORE = 'metricStore',
  NAMESPACE = 'namespace',
  TRACE_STORE = 'traceStore'
}

export enum MeshNodeType {
  BOX = 'box',
  INFRA = 'infra'
}

// TODO
export type MeshNodeHealthData = string;

// Node data expected from server
export interface MeshNodeData {
  // required
  cluster: string;
  id: string;
  infraName: string;
  infraType: MeshInfraType;
  namespace: string;
  nodeType: MeshNodeType;

  // optional
  healthData?: MeshNodeHealthData;
  infraData?: MeshCluster; // add other type options as the case arises
  isAmbient?: boolean;
  isBox?: string;
  isInaccessible?: boolean;
  isMTLS?: boolean;
  isOutOfMesh?: boolean;
  labels?: { [key: string]: string };
  parent?: string;
}

// Edge data expected from server
export interface MeshEdgeData {
  id: string;
  source: string;
  target: string;
}

export interface MeshNodeWrapper {
  data: MeshNodeData;
}

export interface MeshEdgeWrapper {
  data: MeshEdgeData;
}

export interface MeshElements {
  edges?: MeshEdgeWrapper[];
  nodes?: MeshNodeWrapper[];
}

// TODO: unnecessary?
export interface MeshQuery {
  appenders?: AppenderString;
}

export interface MeshDefinition {
  elements: MeshElements;
  timestamp: number;
}

// Node data after decorating at fetch-time (what is mainly used by ui code)
export interface DecoratedMeshNodeData extends MeshNodeData {
  healthStatus: string; // status name
}

// Edge data after decorating at fetch-time (what is mainly used by ui code)
export interface DecoratedMeshEdgeData extends MeshEdgeData {
  // Default value -1
  isMTLS: number;

  // assigned when graph is updated, the edge health depends on the node health, traffic, and config
  healthStatus?: string; // status name
}

export interface DecoratedMeshNodeWrapper {
  data: DecoratedMeshNodeData;
}

export interface DecoratedMeshEdgeWrapper {
  data: DecoratedMeshEdgeData;
}

export interface DecoratedMeshElements {
  edges?: DecoratedMeshEdgeWrapper[];
  nodes?: DecoratedMeshNodeWrapper[];
}

export type MeshType = 'mesh' | 'node' | 'edge' | 'box';

export interface MeshTarget {
  elem: Controller | GraphElement<ElementModel, any> | undefined;
  type: MeshType; // the element type
}

export const MeshAttr = {
  // shared attrs
  id: 'id',

  // edge attrs
  healthStatus: 'healthStatus',
  isMTLS: 'isMTLS',

  // node attrs
  cluster: 'cluster',
  health: 'health',
  infraData: 'infraData',
  infraName: 'infraName',
  infraType: 'infraType',
  isInaccessible: 'isInaccessible',
  isOutOfMesh: 'isOutOfMesh',
  namespace: 'namespace',
  nodeType: 'nodeType'
};
