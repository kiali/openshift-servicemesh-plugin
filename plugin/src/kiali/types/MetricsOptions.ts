import { TargetKind } from './Common';

export interface MetricsQuery {
  avg?: boolean;
  byLabels?: string[];
  duration?: number;
  quantiles?: string[];
  queryTime?: number;
  rateFunc?: string;
  rateInterval?: string;
  step?: number;
}

export interface DashboardQuery extends MetricsQuery {
  additionalLabels?: string;
  labelsFilters?: string;
  rawDataAggregator?: Aggregator;
  workload?: string;
  workloadType?: string;
}

export type Aggregator = 'sum' | 'avg' | 'min' | 'max' | 'stddev' | 'stdvar';

export interface IstioMetricsOptions extends MetricsQuery {
  direction: Direction;
  includeAmbient: boolean;
  filters?: string[];
  requestProtocol?: string;
  reporter: Reporter;
}

export type Reporter = 'source' | 'destination' | 'both';
export type Direction = 'inbound' | 'outbound';

export interface Target {
  cluster?: string;
  kind: TargetKind;
  name: string;
  namespace: string;
}

export interface MetricsStatsQuery {
  avg: boolean;
  direction: Direction;
  interval: string;
  peerTarget?: Target;
  quantiles: string[];
  queryTime: number;
  target: Target;
}

export const statsQueryToKey = (q: MetricsStatsQuery): string =>
  genStatsKey(q.target, q.peerTarget, q.direction, q.interval);

// !! genStatsKey HAS to mirror backend's models.MetricsStatsQuery#GenKey in models/metrics.go
export const genStatsKey = (target: Target, peer: Target | undefined, direction: string, interval: string): string => {
  const peerKey = peer ? genTargetKey(peer) : '';
  return `${genTargetKey(target)}:${peerKey}:${direction}:${interval}`;
};

const genTargetKey = (target: Target): string => {
  return `${target.namespace}:${target.kind}:${target.name}`;
};
