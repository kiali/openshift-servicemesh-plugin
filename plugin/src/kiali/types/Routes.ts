import * as React from 'react';

export interface MenuItem {
  id: string;
  title: string;
  to: string;
  pathsActive?: RegExp[];
}

export interface Path {
  path: string;
  component?: any;
  render?: () => React.ReactNode;
}
