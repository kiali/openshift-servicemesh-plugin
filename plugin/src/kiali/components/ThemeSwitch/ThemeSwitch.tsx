import React from 'react';
import { Switch } from '@patternfly/react-core';
import { KialiAppState } from 'store/Store';
import { connect } from 'react-redux';
import { KIALI_THEME, PF_THEME_DARK, Theme } from 'types/Common';
import { GlobalActions } from 'actions/GlobalActions';
import { store } from 'store/ConfigStore';

type ThemeSwitchProps = {
  theme: string;
};

export const ThemeSwitchComponent = (props: ThemeSwitchProps) => {
  const handleTheme = () => {
    if (props.theme === Theme.LIGHT) {
      document.documentElement.classList.add(PF_THEME_DARK);
      store.dispatch(GlobalActions.setTheme(Theme.DARK));
      localStorage.setItem(KIALI_THEME, Theme.DARK);
    } else {
      document.documentElement.classList.remove(PF_THEME_DARK);
      store.dispatch(GlobalActions.setTheme(Theme.LIGHT));
      localStorage.setItem(KIALI_THEME, Theme.LIGHT);
    }

    // Refresh page to load new theme (certain components are not reloaded like cytoscape graph)
    const refreshTick = new CustomEvent('refreshTick', { detail: Date.now() });
    document.dispatchEvent(refreshTick);
  };

  return (
    <Switch
      id="theme-switch"
      label={'Dark Theme'}
      labelOff={'Dark Theme'}
      isChecked={props.theme === Theme.DARK}
      onChange={handleTheme}
    />
  );
};

const mapStateToProps = (state: KialiAppState) => {
  return {
    theme: state.globalState.theme
  };
};

export const ThemeSwitch = connect(mapStateToProps)(ThemeSwitchComponent);
