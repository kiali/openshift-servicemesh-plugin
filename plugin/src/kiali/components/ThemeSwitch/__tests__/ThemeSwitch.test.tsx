import * as React from 'react';
import { shallowToJson } from 'enzyme-to-json';
import { mount, shallow } from 'enzyme';
import { ThemeSwitchComponent } from '../ThemeSwitch';
import { KIALI_THEME, PF_THEME_DARK, Theme } from 'types/Common';
import { Button } from '@patternfly/react-core';

describe('ThemeSwitch renders', () => {
  it('light theme', () => {
    const wrapper = shallow(<ThemeSwitchComponent theme={Theme.LIGHT} />);

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('dark theme', () => {
    const wrapper = shallow(<ThemeSwitchComponent theme={Theme.DARK} />);

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});

describe('ThemeSwitch changes', () => {
  it('to dark theme', () => {
    const wrapper = mount(<ThemeSwitchComponent theme={Theme.LIGHT} />);
    var buttonLight = () => wrapper.find(Button).at(0);
    buttonLight().simulate('click');
    expect(document.documentElement.classList.contains(PF_THEME_DARK)).toBe(true);
    expect(window.localStorage.getItem(KIALI_THEME)).toBe(Theme.DARK);
  });

  it('to light theme', () => {
    const wrapper = mount(<ThemeSwitchComponent theme={Theme.DARK} />);
    var buttonDark = () => wrapper.find(Button).at(1);
    buttonDark().simulate('click');
    expect(document.documentElement.classList.contains(PF_THEME_DARK)).toBe(false);
    expect(window.localStorage.getItem(KIALI_THEME)).toBe(Theme.LIGHT);
  });
});
