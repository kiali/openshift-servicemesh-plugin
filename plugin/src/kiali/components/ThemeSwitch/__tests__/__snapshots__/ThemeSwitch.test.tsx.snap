// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitch renders dark theme 1`] = `
<Tooltip
  content={
    <React.Fragment>
      Switch to Light Mode
    </React.Fragment>
  }
  position="bottom"
>
  <div>
    <Button
      className="kiali_fu6nkss kiali_fj999mp"
      onClick={[Function]}
      variant="secondary"
    >
      <Sun
        className="kiali_ft7uynw"
      />
    </Button>
    <Button
      className="kiali_fu6nkss kiali_fxx0nx6"
      onClick={[Function]}
      variant="primary"
    >
      <Moon
        className="kiali_ft7uynw"
      />
    </Button>
  </div>
</Tooltip>
`;

exports[`ThemeSwitch renders light theme 1`] = `
<Tooltip
  content={
    <React.Fragment>
      Switch to Dark Mode
    </React.Fragment>
  }
  position="bottom"
>
  <div>
    <Button
      className="kiali_fu6nkss kiali_fj999mp"
      onClick={[Function]}
      variant="primary"
    >
      <Sun
        className="kiali_ft7uynw"
      />
    </Button>
    <Button
      className="kiali_fu6nkss kiali_fxx0nx6"
      onClick={[Function]}
      variant="secondary"
    >
      <Moon
        className="kiali_ft7uynw"
      />
    </Button>
  </div>
</Tooltip>
`;
